marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/provenance/dh-provenance.xqy

```diff
@@ -0,0 +1,128 @@
+declare private variable $user-role as xs:string := "ps-user";
+declare private variable $internal-role as xs:string := "ps-internal";
+declare private variable $ps-dir as xs:string := "/provenance/";
+
+declare private function map-get($map as map:map, $key as xs:string){
+  let $items := map:get($map,$key)
+  return
+    if ($items instance of json:array) then
+      json:array-values($items)
+    else
+      $items
+};
+
+declare private function record-uri($provID as xs:string) as xs:string {
+  let $uri := $ps-dir||xdmp:sha256($provID)||".xml"
+  return $uri
+};
+
+declare private function wrap-in-prov-document($nodes as node()*) {
+  element prov:document {
+    namespace prov {"http://www.w3.org/ns/prov#"},
+    namespace xsi {"http://www.w3.org/2001/XMLSchema-instance"},
+    namespace ps  {"http://marklogic.com/provenance-services"},
+    namespace dh  {"http://marklogic.com/data-hub/prov#"},
+    namespace job  {"http://marklogic.com/data-hub/job#"},
+    namespace step  {"http://marklogic.com/data-hub/step#"},
+    $nodes
+  }
+};
+
+
+declare private function build-components($map as map:map) as node()*{
+  (
+    map:get($map,"activity"),
+    map:get($map,"agent"),
+    map:get($map,"wasAssociatedWith")
+  )
+};
+
+declare private function build($map as map:map) as node(){
+  wrap-in-prov-document(build-components($map))
+};
+
+declare private function job-record-activity($map as map:map, $id as xs:string, $options as map:map*) as map:map {
+  let $prov-id := fn:concat("job:", $id)
+  let $rec :=
+    element { xs:QName("prov:activity") } {
+      attribute prov:id { $prov-id },
+      element prov:label { $id },
+      element prov:type { attribute xsi:type {"xsd:QName"}, "dh:Job" },
+      element prov:startTime { map-get($options,"startTime") }
+    }
+  let $_ := map:put($map, "activity", $rec)
+  return $map
+};
+
+declare private function job-record-agent($map as map:map, $options as map:map*) as map:map {
+  let $user := map-get($options,"user")
+  let $prov-id := fn:concat("user:", $user)
+  let $rec :=
+    element { xs:QName("prov:agent") } {
+      attribute prov:id { $prov-id },
+      element prov:label { $user },
+      element prov:type { attribute xsi:type {"xsd:QName"}, "dh:User" }
+    }
+  let $_ := map:put($map, "agent", $rec)
+  return $map
+};
+
+declare private function job-record-associatedWith($map as map:map, $id as xs:string, $options as map:map*) as map:map {
+  let $user := fn:concat("user:",map-get($options,"user"))
+  let $prov-id := fn:concat("job:", $id)
+  let $rec :=
+    element { xs:QName("prov:wasAssociatedWith") } {
+      element prov:activity { attribute prov:ref {$prov-id} },
+      element prov:agent { attribute prov:ref {$user} }
+    }
+  let $_ := map:put($map, "wasAssociatedWith", $rec)
+  return $map
+};
+
+declare function update-endTime-in-provenance-record($uri as xs:string, $options as map:map*, $target-database as xs:string) as empty-sequence() {
```

Comment:
Let's put public functions at top and private at the bottom, which is consistent with languages like Java. Makes it easier to understand the public API of this module. And generally, that improves readability - i.e. a user can read in a top-down fashion, starting with a public function and then scrolling down to find the private functions called by it.